unison-core/src/Unison/Names3.hs

```diff
@@ -91,6 +78,17 @@ push n1 ns = Names (unionLeft0 n1 cur) (oldNames ns <> shadowed) where
   unionLeft0 n1 n2 = names0 terms' types' where
     terms' = terms0 n1 <> R.subtractDom (R.dom $ terms0 n1) (terms0 n2)
     types' = types0 n1 <> R.subtractDom (R.dom $ types0 n1) (types0 n2)
+  -- For all names in `ns`, (ex: foo.bar.baz), generate the list of suffixes
+  -- of that name [[foo.bar.baz], [bar.baz], [baz]]. Any suffix which uniquely
+  -- refers to a single definition is added as an alias
+  suffixify0 :: Names0 -> Names0
+  suffixify0 ns = ns <> suffixNs
+    where
+    suffixNs = names0 (R.fromList uniqueTerms) (R.fromList uniqueTypes)
+    terms = List.multimap [ (n,ref) | (n0,ref) <- R.toList (terms0 ns), n <- Name.suffixes n0 ]
+    types = List.multimap [ (n,ref) | (n0,ref) <- R.toList (types0 ns), n <- Name.suffixes n0 ]
+    uniqueTerms = [ (n,ref) | (n, nubOrd -> [ref]) <- Map.toList terms ]
+    uniqueTypes = [ (n,ref) | (n, nubOrd -> [ref]) <- Map.toList types ]
```

Comment:
This function `push` was the one spot where we still use suffixification, so it's in a where clause of that function now. I think the names index could be more like a list of indexes which are consulted in turn (kind of like old names, but that also controls whether the identifier is hash qualified or not), and then this function `push` could have a direct implementation without doing suffixification.

Reply:
Go on